### ISSUE TITLE
Support permessage-deflate

### ISSUE THREAD

**Issue by durka** [cyderize/rust-websocket#123]
I think the bottleneck in my app is pushing data over the websocket, and maybe compression would help. I see this item in the roadmap, but I'm wondering how hard it is to add / how far off it might be?

**Comment by illegalprime** [comment]
Thanks for the bug! Having a tracking issue for this is important.

It's not very difficult, you just add the header in the handshake that signals a support for it, then you add a compression / decompression pass when you send / receive each message.
Easy crate for it is: `libz-sys`

As for roadmap both async and this are important, any PRs would be welcome. I'll think about moving this up in priority above async stuff. It's about demand and my time.

What kind of data are you pushing?

**Comment by durka** [comment]
PNGs, which I suspect are already as compressed as they're going to get, and some textual numeric data which I think is pretty compressible.

**Comment by durka** [comment]
Besides those two steps you listed, it might be useful to expose the "permessage" part in the API somehow.

**Comment by illegalprime** [comment]
@durka yeah the plan is to make compression a feature (so it's an optional dependency) then add method in the client and server builders to create connections that support it or not.

For your case, the best compression is when you know the most about the data you are compressing. I wouldn't try to compress the PNGs (although maybe I would put them into a really lossy JPEG), but your metadata can probably be best compressed by your code. So maybe you can encode your numerical metadata in it's binary representation and send that over? (I think there's a serde thing that will convert your struct into binary)

Are you interested in contributing?

**Comment by illegalprime** [comment]
Actually I would recommend `bincode`: https://github.com/TyOverby/bincode

**Comment by durka** [comment]
But for the most control it should be toggle-able by message, not just by connection, right?

Yes I am interested in helping out with this! But I may not have time for a couple of weeks.

**Comment by illegalprime** [comment]
Yup you're absolutely right, I want to put a priority on flexibility. And awesome I'll keep you posted on my progress too.

**Comment by nikolaydimitrov** [comment]
I think I need to wait a couple of years more before starting using Rust :)

**Comment by illegalprime** [comment]
@daweth I was still in school when I was working on this. A lot has changed and I'm not involved anymore.
